extensions/register-and-verify-3.3.md

```diff
@@ -0,0 +1,165 @@
+layout: spec
+copyrights:
+  -
+    name: "William Pitcock"
+    period: "2014-2015"
+    email: "nenolod@dereferenced.org"
+---
+
+The `REGISTER` and `VERIFY` commands provide a unified, standardized interface for
+account creation and validation to the network's authentication layer.  This allows for a
+network to provide a common interface regardless of what authentication layer they have chosen
+for their network to operate, i.e. traditional services, a different central authority, or a
+decentralized model similar to [IRCX][ircx].
+
+   [ircx]: http://en.wikipedia.org/wiki/IRCX
+   
+We believe that the benefit of adding a common interface for account creation and validation,
+is helpful to the end user in discovering the capabilities offered by the authentication and
+authority layer of the network.  Further, we believe it is helpful to client authors, so that
+they may provide guided account creation and validation for their users, regardless of network
+configuration.
+
+## The `REGISTER` command
+
+The `REGISTER` command signals the intent of a client to register an account in the network's
+authentication layer.  It is similar to current methods of signalling that intent.
+
+A `REGISTER` command consists of the following format:
+
+`REGISTER <accountname> <callback> :<passphrase>`
+
+A passphrase MAY have spaces in it.
+
+The callback field designates an opaque value which indicates where a verification code, if any,
+will be sent.  The callback field is implementation-specific, with namespaces being listed using the
+`REGCALLBACKS` ISUPPORT token.  An invalid callback parameter should result in the `ERR_REG_INVALID_CALLBACK`
+error.  In the event that a callback is not provided, the client SHOULD send `*` to indicate that
+they are not providing a callback resource.
+
+The IRC server MAY forward the `REGISTER` command to a central authority, or process it locally, however
+the authentication layer SHOULD NOT send any legacy messages in reply to a `REGISTER` command.
+
+Upon success, the IRC server MUST send the `RPL_REGISTRATIONSUCCESS` numeric, which looks like:
+
+    :<server> 920 <user_nickname> <accountname> :Account registered
+
+If the server requires a verification token, it must also reply with the `RPL_REG_VERIFICATION_REQUIRED`
+numeric, which looks like:
+
+    :<server> 927 <user_nickname> <accountname> <callback> :A verification token was sent
+
+Upon error, the IRC server MUST send an error code that is relevant.  We suggest these numerics:
+
+`ERR_ACCOUNT_ALREADY_EXISTS`:
+    `:<server> 921 <user_nickname> <accountname> :Account already exists`
+`ERR_REG_INVALID_PARAMS`:
+    `:<server> 922 <user_nickname> <accountname> :Invalid params: <descriptive_text>`
+`ERR_REG_INVALID_CALLBACK`:
+    `:<server> 926 <user_nickname> <accountname> <callback> :Callback token is invalid`
+
+The server MAY send additional informative text upon registration success or failure using `RPL_TEXT` or `NOTICE`.
```

Comment:
This could warrant a numeric itself. There is no way to tell what services may require or warn about, so for clients processing these it would be handy to detect any arbitrary text/warnings relating to registration.